### PR TITLE
feat: Introduce a SignerPool with a NonceManager

### PR DESCRIPTION
**Pull request by davidvuong**

### CODE REVIEW

File: src/distributor.ts [hunk 1]
```diff
@@ -142,7 +144,7 @@ export class Distributor {
         } catch (err) {
           this.logger.error('Encountered error at distributor loop', { args: { err } });
         }
-        await this.delay(this.distributorProcessInterval);
+        await delay(this.distributorProcessInterval);
```
Comment (0xjocke) [inline src/distributor.ts:148]: Can remove the delay method in this class now. Line 33

File: src/signerpool.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    this.pool = Array.from(Array(this.signers.length).keys());
+    this.logger = createLogger('SignerPool');
+  }
+
+  private async acquire(ctx: WithSignerContext): Promise<[number, NonceManager]> {
+    this.logger.info(`[${ctx.asset}] Awaiting signer`);
+    let i = this.pool.pop();
+
+    while (i === undefined) {
+      delay(10);
```
Comment (0xjocke) [inline src/signerpool.ts]: Missing await

File: src/signerpool.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    } catch (err) {
+      if (isObjectOrErrorWithCode(err)) {
+        // Special handeling for NONCE_EXPIRED
+        if (err.code === 'NONCE_EXPIRED') {
+          this.logger.error(err.toString());
+          const nonce = signer.getTransactionCount('latest');
+          this.logger.info(`[${ctx.asset}] Updating nonce for Nonce manager to nonce: '${nonce}'`);
+          signer.setTransactionCount(nonce);
+        }
+      }
```
Comment (0xjocke) [inline src/signerpool.ts:90]: I think we should rethrow the error here, assuming consumers catching it? 
If not at least log it?

**Comment by davidvuong** [comment]
@bachstatter Good catch on everything! Fixed.

**Review by 0xjocke — APPROVED** [review]

**Review by leomassazza — APPROVED** [review]
LGTM

**Pull request merged**